### PR TITLE
Update Doctrine components and add Symfony 5.2 support

### PR DESCRIPTION
**Pull request by guillaume-ro-fr**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | yes
| New feature?      | yes
| BC breaks?        | yes     
| Need Doc update   | yes

## Describe your change

- Add Symfony 5.2 support (requires `symfony/doctrine-bridge:5.2`-> `doctrine/persistence:^2`)
- Migrate from `doctrine/common` (deprecated) to `doctrine/*` components
- Dropping support for PHP < 7.1 (`doctrine/event-manager` and `doctrine/persistence`) (EOL for PHP 7.1 was in 2019)

## What problem is this fixing?

- Symfony 5.2 support

### CODE REVIEW

**Comment by guillaume-ro-fr** [comment]
I don't know why, but environment variables for `ALGOLIA_APP_ID` and `ALGOLIA_API_KEY` are not set in CI.
Results on my local environment with variables:

```
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.13 with Xdebug 2.9.6
Configuration: /redacted/search-bundle/phpunit.xml.dist

......................................                            38 / 38 (100%)

Time: 3.22 minutes, Memory: 26.00 MB

OK (38 tests, 155 assertions)

Generating code coverage report in Clover XML format ... done
```

**Comment by sylfabre** [comment]
Thank you @guillaume-ro-fr, this will fix https://github.com/algolia/search-bundle/issues/356

**Comment by chloelbn** [comment]
Hi @guillaume-ro-fr ! Would you mind rebasing and pushing your PR again please? Thanks!

**Comment by guillaume-ro-fr** [comment]
Hi @chloelbn ! I fixed some tests (my cache was outdated in local), but the Algolia client ID is unavailable in Circle CI tests : `Environment variable not found: "ALGOLIA_APP_ID"` and `Impossible to connect, please check your Algolia Application Id.`.

**Comment by chloelbn** [comment]
Hi @guillaume-ro-fr, indeed I thought my previous commit would fix that. I'm investigating and will keep you posted

**Comment by sylfabre** [comment]
Hello @chloelbn
When can we expect this PR to be merged please?

**Comment by chloelbn** [comment]
Hi @sylfabre and @guillaume-ro-fr ! sorry for the delay. I'd need you to rebase again please, a fix was done for the CI

**Comment by guillaume-ro-fr** [comment]
Hi @chloelbn ! I don't know why, but there are always 1 or 2 tests on 4 that failed (randomly between 7.1 ... 7.4, 5.4 and 7.0 failed on `composer install` because of `doctrine/persistance` as explain in my first comment).

```
#!/bin/bash -eo pipefail
if [ "$CIRCLE_PR_REPONAME" ]; then
  curl -s https://algoliasearch-client-keygen.herokuapp.com | sh >> $BASH_ENV
fi
sh: 2: Syntax error: newline unexpected

Exited with code exit status 2
CircleCI received exit code 2
```

API Key not ready for the first test (race condition?)
```
1) Algolia\SearchBundle\TestCase\AggregatorTest::testAggregatorProxyClass
Algolia\AlgoliaSearch\Exceptions\BadRequestException: Invalid Application-ID or API key
```

**Comment by chloelbn** [comment]
Hi @guillaume-ro-fr ! Indeed sometimes it fails because of a race condition but we also get this error that concerns me more

```
TypeError: Argument 2 passed to JMS\Serializer\Metadata\Driver\AbstractDoctrineTypeDriver::__construct() must be an instance of Doctrine\Common\Persistence\ManagerRegistry, instance of Doctrine\Bundle\DoctrineBundle\Registry given, called in /home/circleci/project/var/cache/test/ContainerIMgqkUK/getJmsSerializer_MetadataDriverService.php on line 11
```

**Comment by guillaume-ro-fr** [comment]
Hi @chloelbn ! I just found out that `jms/serializer-bundle` 2.4 is not compatible with `doctrine/persistence` and `jms/serializer-bundle` 3.0 is not compatible with PHP < 7.2.

Maybe we can also drop support for PHP < 7.2 (PHP 7.1 is deprecated since 2019)?

PHP supported version by `algolia/search-bundle` : 
- 7.2 (EOL december 2020)
- 7.3
- 7.4
- 8.0

**Comment by sylfabre** [comment]
@chloelbn Do you have some news about this one please?

**Comment by chloelbn** [comment]
@sylfabre @guillaume-ro-fr yes, I think we'll have to release a new major to get our dependencies up-to-date soon! I'll work on it asap

**Review by matpo — APPROVED** [review]

**Comment by tristanbes** [comment]
👀

**Comment by guillaume-ro-fr** [comment]
Tests re-failed because of the race condition with the API key broker.

**Comment by chloelbn** [comment]
Could you also update phpunit dependency please?

**Comment by guillaume-ro-fr** [comment]
@chloelbn Is there any reason why this test is performed 10 times? (Confirming that Algolia servers don't change the value (???) or this an error?)

https://github.com/algolia/search-bundle/blob/a26bf6eb79e747b8ca197e4e5d3c3703d271256d/tests/TestCase/CommandsTest.php#L113-L117

We can gain at least 1 min 40 on the tests if it's a mistake (3*10*1 sec + 2*5*1 sec)

**Comment by chloelbn** [comment]
@guillaume-ro-fr this ensures there's no flakiness with the command taking some time

**Comment by samijnih** [comment]
👀 😂

File: tests/TestCase/EntityProxyTest.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        // Unset env variables to make sure Algolia
+        // Credentials are only required when the
+        // client is used. Save them to restore them after.
+        // See: https://github.com/algolia/search-bundle/issues/241
+        self::$values = [
+            'env_id'  => getenv('ALGOLIA_APP_ID'),
+            'env_key' => getenv('ALGOLIA_API_KEY'),
+            '_env'    => $_ENV,
+            '_server' => $_SERVER,
```
Comment (DevinCodes) [inline tests/TestCase/EntityProxyTest.php:23]: Can we store only the `$_ENV` and `$_SERVER` keys that we unset, please? It doesn't seem super safe to me to keep the entire array in memory when we don't have to.

File: src/Util/ClassInfo.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     *
+     * @param string $className
+     *
+     * @return string
+     */
+    public static function getRealClassName($className)
+    {
+        // Define variable for static analysis
+        $positionPm = false;
+        // __CG__: Doctrine Common Marker for Proxy (ODM < 2.0 and ORM < 3.0)
```
Comment (DevinCodes) [inline src/Util/ClassInfo.php:35]: I've seen a test case for the case below in `tests/TestCase/EntityProxyTest.php`, but we're missing a test for the `__CG__` case as far as I can see. Could you add this, please?

**Review by DevinCodes — CHANGES_REQUESTED** [review]
Thank you @guillaume-ro-fr for your work on this, this is amazing! I just have 2 comments, but other than that I think it's good to go. Let me know what you think 😄

**Comment by guillaume-ro-fr** [comment]
Tests (PHP 8.0) re-failed because of the race condition with the API key broker.

`/tmp/.bash_env-6050c10bb1001030eb79cae8-0-build: line 1: syntax error near unexpected token 'newline'
/tmp/.bash_env-6050c10bb1001030eb79cae8-0-build: line 1: '<!DOCTYPE html>'`

**Review by DevinCodes — APPROVED** [review]
Thank you, this looks good to go to me! 😄 

@chloelbn could you please take a look at the tests? Thank you in advance!

**Comment by chloelbn** [comment]
Hi @DevinCodes ! The code are failing because of the API key dealer, let's handle that directly on its own repository as it's impacting all of us. Thanks for your review! And thanks for the updates too @guillaume-ro-fr ! We'll soon be able to roll

**Review by chloelbn — APPROVED** [review]

**Comment by chloelbn** [comment]
@guillaume-ro-fr thanks to you, v5 is out!!! Thanks a TON for your work and helping this compatibility happen. We really appreciate it 🔥 you rock!

**Pull request merged**